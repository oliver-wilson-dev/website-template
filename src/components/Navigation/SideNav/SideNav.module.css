@import "../../../styles/colours.css";

.icon {
  position: absolute;
  height: 30px;
  right: 25px;
}

.burgerMenuHide {
  display: none;
}

.crossIcon {
  fill: black;
  opacity: 0;
  z-index: 10;

  transition-property: opacity;
  transition-duration: 0.3s;
  transition-timing-function: ease;
}

.crossIconShow {
  opacity: 1;
}

.menuContainer {
  background-color: transparent;

  height: 100vh;
  z-index: 5;
  position: absolute;
  top: 0;
  right: 0;

  display: flex;
  justify-content: flex-end;
  width: 100vw;
}

@keyframes open {
  0% {
    transform: translate(100vw);
    background-color: transparent;
  }

  50% {
    transform: translate(0);
    background-color: transparent;
  }

  100% {
    transform: translate(0);
    background: rgba(0,0,0,.288);
  }
}

@keyframes close {
  0% {
    transform: translate(0);
    background: rgba(0,0,0,.288);
  }

  50% {
    transform: translate(0);
    background-color: transparent;
  }

  100% {
    transform: translate(100vw);
    background-color: transparent;
  }
}

.menuOpen {
  animation-name: open;
  animation-duration: 0.5s;
  animation-fill-mode: forwards;
  animation-direction: normal;
}

.menuClose {
  animation-name: close;
  animation-duration: 0.5s;
  animation-fill-mode: forwards;
  animation-direction: normal;
}

.menuHideFirstRender {
  display: none;
}

.menuContent {
  width: 50%;
  height: 100%;
  padding: 50px 0px;
  background-color: $background-grey;
  box-shadow: 0px 0px 8px black;

  display: flex;
  flex-direction: column;
  justify-content: ;
  opacity: 1;
}

.menuList {
  width: 100%;
  padding: 0px;
  margin: 20px 0px;
  list-style-type: none;
  display: flex;
  flex-direction: column;
  max-height: 80%;
  overflow-y: scroll;
  overflow-x: hidden;
  justify-content: flex-start;
}

.menuItem {
  text-align: center;
  font-weight: 700;
  font-size: 24px;
  font-family: "Source Sans Pro", sans-serif;

  display: flex;
  justify-content: center;
  align-items: center;
  cursor: pointer;

  &:hover,
  &:focus {
    background-color: $dark-grey;

    & .link {
      color: $text-light-grey;
    }
  }
}

.menuItemMargin {
  margin: 20px 25px 20px 20px;
}

.link {
  color: $text-dark-grey;
  width: 100%;
  height: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
}

@media (min-width: 768px) {
  .menuContainer, .icon, .crossIcon {
    display: none;
  }
}
