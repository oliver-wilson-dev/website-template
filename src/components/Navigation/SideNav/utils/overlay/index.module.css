.bodyFixed {
  position: fixed;
  left: 0;
  width: 100%;
  overflow-y: scroll;
}
