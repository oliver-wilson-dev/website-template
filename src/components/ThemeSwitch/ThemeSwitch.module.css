.moon {
    width: 30px;
    height: 30px;
}

.sun {
    width: 30px;
    height: 30px;
}

.btn {
    background-color: unset;
    border: none;
    padding: 0px;
    opacity: 1;

    transition: transform 0.1s;
    transition-timing-function: linear;
}

.hide {
    transform: scale(0);
}

.show {
    transform: scale(1);
}