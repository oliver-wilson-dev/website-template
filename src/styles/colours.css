$background-grey: #e9ecef;
$dark-grey:  #333333;
$text-dark-grey:  #333333;
$text-light-grey:  #ffffff;
