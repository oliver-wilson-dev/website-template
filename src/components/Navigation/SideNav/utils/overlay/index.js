import styles from './index.module.css';

export const preventScroll = () => {
  document.body.style.top = `-${window.scrollY}px`;
  document.body.classList.add(styles.bodyFixed);
};

export const allowScroll = () => {
  const scrollY = document.body.style.top;
  document.body.classList.remove(styles.bodyFixed);
  document.body.style.top = '';
  window.scrollTo(0, parseInt(scrollY || '0', 10) * -1);
};
