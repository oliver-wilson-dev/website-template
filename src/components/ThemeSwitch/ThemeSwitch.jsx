import React, { useState, useEffect } from 'react';
import cn from 'classnames';
import Moon from './assets/moon.svg';
import Sun from './assets/sun.svg';
import styles from './ThemeSwitch.module.css';
import { LIGHT_THEME } from '../../styles/constants';

const ThemeSwitch = ({ theme, toggleTheme, className }) => {
  const [hide, setHide] = useState(false);
  const [hideComplete, setHideComplete] = useState(false);

  useEffect(() => {
    if (hide && hideComplete) {
      setHide(false);
      setHideComplete(false);
      toggleTheme();
    }
  }, [theme, hide, hideComplete]);

  const onClick = () => {
    setHide(true);
  };

  const onTransitionEnd = () => {
    if (hide && !hideComplete) {
      setHideComplete(true);
    }
  };

  return (
    <button className={cn(styles.btn, { [styles.hide]: hide, [styles.show]: !hide }, className)} type="button" onClick={onClick} onTransitionEnd={onTransitionEnd}>
      {theme === LIGHT_THEME ? <Moon className={styles.moon} /> : <Sun className={styles.sun} />}
    </button>
  );
};

export default ThemeSwitch;
