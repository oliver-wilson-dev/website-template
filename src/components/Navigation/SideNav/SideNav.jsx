/* eslint-disable jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions */
import React, { useEffect, useRef, useState } from 'react';
import cn from 'classnames';
import { NavLink } from 'react-router-dom';
import CrossIcon from './assets/cross.svg';
import BurgerMenu from './assets/burgerMenu.svg';
import { preventScroll, allowScroll } from './utils/overlay';
import routes from '../../../routes';
import styles from './SideNav.module.css';

const { home, about, contact } = routes;

const ignoreClick = (event) => {
  event.stopPropagation();
};

const SideNav = () => {
  const isFirstRender = useRef(true);
  const [showSideNav, setShowSideNav] = useState(false);

  const toggleShowSideNav = () => {
    if (isFirstRender.current) {
      isFirstRender.current = false;
    }
    setShowSideNav(!showSideNav);
  };

  useEffect(() => {
    if (showSideNav) {
      preventScroll();
    } else {
      allowScroll();
    }
  }, [showSideNav]);

  const onOverlayClick = () => {
    toggleShowSideNav();
  };

  return (
    <>
      <BurgerMenu
        className={cn(styles.icon, {
          [styles.burgerMenuHide]: showSideNav
        })}
        onClick={toggleShowSideNav}
      />
      <CrossIcon
        className={cn(styles.crossIcon, styles.icon, {
          [styles.crossIconShow]: showSideNav
        })}
        onClick={toggleShowSideNav}
      />
      <div
        className={cn(styles.menuContainer, {
          [styles.menuOpen]: showSideNav,
          [styles.menuClose]: !showSideNav,
          [styles.menuHideFirstRender]: isFirstRender.current
        })}
        onClick={onOverlayClick}
      >
        <div
          onClick={ignoreClick}
          className={styles.menuContent}
        >
          <ul className={styles.menuList}>
            <li className={styles.menuItem}>
              <NavLink
                to={home.route}
                className={cn(styles.link, styles.menuItemMargin)}
                onClick={toggleShowSideNav}
              >
                Home
              </NavLink>
            </li>
            <li className={styles.menuItem}>
              <NavLink
                to={about.route}
                className={cn(styles.link, styles.menuItemMargin)}
                onClick={toggleShowSideNav}
              >
                About
              </NavLink>
            </li>
            <li className={styles.menuItem}>
              <NavLink
                to={contact.route}
                className={cn(styles.link, styles.menuItemMargin)}
                onClick={toggleShowSideNav}
              >
                Contact
              </NavLink>
            </li>
          </ul>
        </div>
      </div>
    </>
  );
};

export default SideNav;
