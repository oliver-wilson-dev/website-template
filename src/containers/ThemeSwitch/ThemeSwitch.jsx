import { connect } from 'react-redux';
import ThemeSwitch from '../../components/ThemeSwitch';
import { toggleTheme } from '../../state/actions';
import { getTheme } from '../../state/selectors';

const mapStateToProps = (state) => ({
  theme: getTheme({ state }),
});

const mapDispatchToProps = (dispatch) => ({
  toggleTheme: () => dispatch(toggleTheme())
});

export default connect(mapStateToProps, mapDispatchToProps)(ThemeSwitch);
