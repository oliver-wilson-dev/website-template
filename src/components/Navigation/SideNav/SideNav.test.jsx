import React from 'react';
import { shallow, mount } from 'enzyme';
import SideNav from './index';
import ThemeSwitch from '../../../containers/ThemeSwitch';
import styles from './SideNav.module.css';
import { preventScroll, allowScroll } from './utils/overlay';

jest.mock('react-router-dom', () => ({ NavLink: ({ children }) => <button type="button">{children}</button> }));

jest.mock('./utils/overlay', () => ({
  preventScroll: jest.fn(),
  allowScroll: jest.fn()
}));

jest.mock('./assets/cross.svg', () => {
  const CrossIcon = (props) => <button type="button" {...props} />;

  return CrossIcon;
});

jest.mock('./assets/burgerMenu.svg', () => {
  const BurgerMenu = (props) => <button type="button" {...props} />;

  return BurgerMenu;
});

jest.mock('../../../containers/ThemeSwitch', () => {
  const ThemeSwitch = (props) => <button type="button" {...props} />;

  return ThemeSwitch;
});

const render = (
  renderMethod = shallow
) => renderMethod(<SideNav />);

describe('<SideNav/> component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should exist', () => {
    expect(render().exists()).toBe(true);
  });

  it('should render correctly', () => {
    expect(render()).toMatchSnapshot();
  });

  describe('when clicking on the burger menu', () => {
    it('should call preventScroll', () => {
      const burgerMenu = render(mount).find(`.${styles.icon}`).first();
      burgerMenu.simulate('click');

      expect(preventScroll).toHaveBeenCalledTimes(1);
    });

    it('should open the side nav', () => {
      const component = render(mount);
      const burgerMenu = component.find(`.${styles.icon}`).first();
      burgerMenu.simulate('click');

      expect(component).toMatchSnapshot();
    });

    describe('and then clicking on the menu content', () => {
      it('should stop the propagation of that event', () => {
        const component = render(mount);
        const burgerMenu = component.find(`.${styles.icon}`).first();
        burgerMenu.simulate('click');

        const menuContent = component.find(`.${styles.menuContent}`).first();
        const mockStopPropagation = jest.fn();
        menuContent.simulate('click', { stopPropagation: mockStopPropagation });

        expect(mockStopPropagation).toHaveBeenCalled();
      });
    });

    describe('and then clicking on the overlay', () => {
      it('should close the side nav', () => {
        const component = render(mount);
        const burgerMenu = component.find(`.${styles.icon}`).first();
        burgerMenu.simulate('click');

        const overlay = component.find(`.${styles.menuContainer}`).first();

        overlay.simulate('click');

        expect(component).toMatchSnapshot();
      });
    });

    describe('and then clicking on the cross icon', () => {
      it('should close the side nav', () => {
        const component = render(mount);
        const burgerMenu = component.find(`.${styles.icon}`).first();
        burgerMenu.simulate('click');

        const crossIcon = component.find(`.${styles.crossIcon}`).first();
        crossIcon.simulate('click');

        expect(component).toMatchSnapshot();
      });

      it('should allow scrolling again', () => {
        const component = render(mount);
        const burgerMenu = component.find(`.${styles.icon}`).first();
        burgerMenu.simulate('click');

        const crossIcon = component.find(`.${styles.crossIcon}`).first();
        crossIcon.simulate('click');

        expect(allowScroll).toHaveBeenCalled();
      });
    });
  });
});
