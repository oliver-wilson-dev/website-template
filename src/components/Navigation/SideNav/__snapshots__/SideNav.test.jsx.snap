// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SideNav/> component should render correctly 1`] = `
<Fragment>
  <BurgerMenu
    className="icon"
    onClick={[Function]}
  />
  <CrossIcon
    className="crossIcon icon"
    onClick={[Function]}
  />
  <div
    className="menuContainer menuClose menuHideFirstRender"
    onClick={[Function]}
  >
    <div
      className="menuContent"
      onClick={[Function]}
    >
      <ul
        className="menuList"
      >
        <li
          className="menuItem"
        >
          <NavLink
            className="link menuItemMargin"
            onClick={[Function]}
            to="/"
          >
            Home
          </NavLink>
        </li>
        <li
          className="menuItem"
        >
          <NavLink
            className="link menuItemMargin"
            onClick={[Function]}
            to="/about"
          >
            About
          </NavLink>
        </li>
        <li
          className="menuItem"
        >
          <NavLink
            className="link menuItemMargin"
            onClick={[Function]}
            to="/contact"
          >
            Contact
          </NavLink>
        </li>
      </ul>
    </div>
  </div>
</Fragment>
`;

exports[`<SideNav/> component when clicking on the burger menu and then clicking on the cross icon should close the side nav 1`] = `
<SideNav>
  <BurgerMenu
    className="icon"
    onClick={[Function]}
  >
    <button
      className="icon"
      onClick={[Function]}
      type="button"
    />
  </BurgerMenu>
  <CrossIcon
    className="crossIcon icon"
    onClick={[Function]}
  >
    <button
      className="crossIcon icon"
      onClick={[Function]}
      type="button"
    />
  </CrossIcon>
  <div
    className="menuContainer menuClose"
    onClick={[Function]}
  >
    <div
      className="menuContent"
      onClick={[Function]}
    >
      <ul
        className="menuList"
      >
        <li
          className="menuItem"
        >
          <NavLink
            className="link menuItemMargin"
            onClick={[Function]}
            to="/"
          >
            <button
              type="button"
            >
              Home
            </button>
          </NavLink>
        </li>
        <li
          className="menuItem"
        >
          <NavLink
            className="link menuItemMargin"
            onClick={[Function]}
            to="/about"
          >
            <button
              type="button"
            >
              About
            </button>
          </NavLink>
        </li>
        <li
          className="menuItem"
        >
          <NavLink
            className="link menuItemMargin"
            onClick={[Function]}
            to="/contact"
          >
            <button
              type="button"
            >
              Contact
            </button>
          </NavLink>
        </li>
      </ul>
    </div>
  </div>
</SideNav>
`;

exports[`<SideNav/> component when clicking on the burger menu and then clicking on the overlay should close the side nav 1`] = `
<SideNav>
  <BurgerMenu
    className="icon"
    onClick={[Function]}
  >
    <button
      className="icon"
      onClick={[Function]}
      type="button"
    />
  </BurgerMenu>
  <CrossIcon
    className="crossIcon icon"
    onClick={[Function]}
  >
    <button
      className="crossIcon icon"
      onClick={[Function]}
      type="button"
    />
  </CrossIcon>
  <div
    className="menuContainer menuClose"
    onClick={[Function]}
  >
    <div
      className="menuContent"
      onClick={[Function]}
    >
      <ul
        className="menuList"
      >
        <li
          className="menuItem"
        >
          <NavLink
            className="link menuItemMargin"
            onClick={[Function]}
            to="/"
          >
            <button
              type="button"
            >
              Home
            </button>
          </NavLink>
        </li>
        <li
          className="menuItem"
        >
          <NavLink
            className="link menuItemMargin"
            onClick={[Function]}
            to="/about"
          >
            <button
              type="button"
            >
              About
            </button>
          </NavLink>
        </li>
        <li
          className="menuItem"
        >
          <NavLink
            className="link menuItemMargin"
            onClick={[Function]}
            to="/contact"
          >
            <button
              type="button"
            >
              Contact
            </button>
          </NavLink>
        </li>
      </ul>
    </div>
  </div>
</SideNav>
`;

exports[`<SideNav/> component when clicking on the burger menu should open the side nav 1`] = `
<SideNav>
  <BurgerMenu
    className="icon burgerMenuHide"
    onClick={[Function]}
  >
    <button
      className="icon burgerMenuHide"
      onClick={[Function]}
      type="button"
    />
  </BurgerMenu>
  <CrossIcon
    className="crossIcon icon crossIconShow"
    onClick={[Function]}
  >
    <button
      className="crossIcon icon crossIconShow"
      onClick={[Function]}
      type="button"
    />
  </CrossIcon>
  <div
    className="menuContainer menuOpen"
    onClick={[Function]}
  >
    <div
      className="menuContent"
      onClick={[Function]}
    >
      <ul
        className="menuList"
      >
        <li
          className="menuItem"
        >
          <NavLink
            className="link menuItemMargin"
            onClick={[Function]}
            to="/"
          >
            <button
              type="button"
            >
              Home
            </button>
          </NavLink>
        </li>
        <li
          className="menuItem"
        >
          <NavLink
            className="link menuItemMargin"
            onClick={[Function]}
            to="/about"
          >
            <button
              type="button"
            >
              About
            </button>
          </NavLink>
        </li>
        <li
          className="menuItem"
        >
          <NavLink
            className="link menuItemMargin"
            onClick={[Function]}
            to="/contact"
          >
            <button
              type="button"
            >
              Contact
            </button>
          </NavLink>
        </li>
      </ul>
    </div>
  </div>
</SideNav>
`;
